### ISSUE TITLE
Rule 058 incorrectly flagging case flattening

### ISSUE THREAD

**Issue by DawidJanczak** [sqlfluff/sqlfluff#3110]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Hello!

I have a nested case statement that unfortunately cannot be flattened, but rule 058 flags it as possible to flatten. It's a very confusing piece of code, but still valid SQL. I simplified this as much as I can:
```
CASE foo
  WHEN 'a' THEN
    TRUE
  ELSE
    bar.date > CURRENT_DATE + CASE foo
      WHEN 'b' THEN
        8
      WHEN 'c' THEN
        9
      END
    AND (...) -- other conditions only triggered when foo is NOT 'a'
END
```

Leaving aside the complexity of this I wonder if rule 058 should only be checking the nested cases when they're not part of another expression?

### Expected Behaviour

I expect the rule not to trigger on nested expressions that cannot be flattened.

### Observed Behaviour

Lint fails with `Nested 'CASE' statement in 'ELSE' clause could be flattened.`

### How to reproduce

Easiest way with hardcoded values:
```
SELECT CASE 'b'
  WHEN 'a' THEN
    TRUE
  ELSE
    '2022-01-01'::date > CURRENT_DATE + CASE 'b'
      WHEN 'b' THEN
        8
      WHEN 'c' THEN
        9
    END
  --  AND (...) -- other conditions only triggered when foo is NOT 'a'
  END AS test
```

### Dialect

Postgres

### Version

sqlfluff, version 0.12.0
Python 3.10.4

### Configuration

```
[sqlfluff]
dialect = postgres

[sqlfluff:rules]
tab_space_size = 2
max_line_length = 120

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = upper

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = upper
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
